Ignore Analytics in the legacy `woocommerce_admin_reports`

tomalec opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Ignore Analytics report entries in the legacy `woocommerce_admin_reports` filter

Fixes https://github.com/woocommerce/woocommerce/issues/36737

Also, filtering the `$reports` array will be helpful to check if there is anything to throw the deprecation warning in our efforts to retire Legacy reports https://github.com/woocommerce/woocommerce/pull/51557

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This is a developer facing fix

1. Enable `WP_DEBUG`
2. Add snippets that mimic an extension adding analytics and legacy reports using the colliding filter `woocommerce_admin_reports`:
```php
// Legacy reports
add_filter( 'woocommerce_admin_reports', function( $reports ) {
    $reports['automatewoo'] = array(
        'title'   => 'AutomateWoo',
        'reports' => array(
            'runs_by_date' => array(
                'title'       => 'Workflow Runs',
                'description' => 'foo',
                'function'    => 'automatewoo_reports_get_runs_by_date',
            ),
            'conversions' => array(
                'title'       => 'Conversions',
                'description' => 'bar',
                'hide_title'  => true,
            ),
        ),
    );
    return $reports;
} );

// Analytics
add_filter( 'woocommerce_admin_reports', function( $reports ) {
    $reports[] = array(
        'slug' => 'aw-workflows',
        'description' => 'AutomateWoo Workflows',
        'url' => '/analytics/aw-workflows',
        'path' => '/automatewoo/v1/workflows',
    );
    return $reports;
} );
```
4. Open the reports page `/wp-admin/admin.php?page=wc-reports&tab=orders`
5. You can also check the analytics endpoint for regression by sending 
```js
fetch("https://consentmode.test/wp-json/wc-analytics/reports", {
  "headers": {
    "x-wp-nonce": "1234567" // nonce you can copy from any other request already bade
  },
});
```

##### Before
![image](https://github.com/user-attachments/assets/62233bd9-dc8e-4377-81dc-2162f4f88bd2)

##### After 
![image](https://github.com/user-attachments/assets/4bcacb2f-d345-4b02-8f9a-6c5579388122)


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

tomalec committed:
Ignore Analytics reports in Legacy reports `woocommerce_admin_reports` filter

Fixes https://github.com/woocommerce/woocommerce/issues/36737

tomalec committed:
Add changelog entry

tomalec committed:
Renamce changelog file

tomalec committed:
add whitespace for PHPCS

mikkamp reviewed (approved):
Thanks for the changes it's working as expected, and if I use the array structure as suggested in the code snippets, it works correctly (invalid entries are skipped). I also tested the REST request and I can see it skip the entries (as was already being done).

However I'd suggest for any hooked functions to do a simple check if the filter should be applied, by checking if it's a REST request or if they are in the admin screen:

```php
// Legacy reports
add_filter(
	'woocommerce_admin_reports',
	function( $reports ) {
		if ( is_admin() ) {
			$reports['automatewoo'] = array(
				'title'   => 'AutomateWoo',
				'reports' => array(
					'runs_by_date' => array(
						'title'       => 'Workflow Runs',
						'description' => 'foo',
						'function'    => 'automatewoo_reports_get_runs_by_date',
					),
					'conversions' => array(
						'title'       => 'Conversions',
						'description' => 'bar',
						'hide_title'  => true,
					),
				),
			);
		}

		return $reports;
	}
);

// Analytics
add_filter(
	'woocommerce_admin_reports',
	function( $reports ) {
		if ( wp_is_serving_rest_request() ) {
			$reports[] = array(
				'slug' => 'aw-workflows',
				'description' => 'AutomateWoo Workflows',
				'url' => '/analytics/aw-workflows',
				'path' => '/automatewoo/v1/workflows',
			);
		}

		return $reports;
	}
);
```

Maybe we should add a docblock for the first filter and then explain the recommended condition checks to add. At least for anyone that discovers it until we fully deprecate.

Since it is working I'll go ahead and approve the PR already.

tomalec committed:
Add explanatory note to `woocommerce_admin_reports` filter

https://github.com/woocommerce/woocommerce/pull/51592#pullrequestreview-2322768343

tomalec commented:
Thanks for the review and great suggestion, added a note in https://github.com/woocommerce/woocommerce/pull/51592/commits/7bf36b5ceda6f7716bb402c191a3b602e491d2b5

tomalec committed:
Merge branch 'trunk' into fix/36737-deprecate-legacy_admin_reports-filter

tomalec committed:
fix phpcs

tomalec commented:
I added the note to the legacy case, as we're planning to eventually remove the legacy code. Therefore, such a check would not be needed in Analytics and could later complicate our guidelines for the new analytics-only implementations.

I added your snippet as a solution to https://github.com/woocommerce/woocommerce/issues/36737.

If you think such a note would be beneficiary for the Analytics filter, we can add it in a new PR :) I'd use code changes in this one, to update deprecations feature branch.
